js/viz.js

```diff
@@ -1,29 +1,111 @@
 // eslint-disable-next-line no-unused-vars
 const viz = {
-  draw(context, websites) {
-    const pi = Math.PI * 2;
-    let x = 50, y = 50;
-
-    context.fillStyle = 'white';
-    context.beginPath();
-
-    for (const website in websites) {
-      context.moveTo(x, y);
-      context.arc(x, y, 10, 0, pi);
-      context.fillText(website, x, y);
-
-      let x1 = x;
-      for (const thirdParty in websites[website].thirdPartyRequests) {
-        x1+=150;
-        context.moveTo(x1, y);
-        context.arc(x1, y, 5, 0, pi);
-        context.fillText(thirdParty, x1, y);
+  draw(nodesData, linksData) {
+    // set up area of the page to put the graph
+    const svg = d3.select('svg');
+    const width = svg.attr('width');
+    const height = svg.attr('height');
+
+    // set up simulation
+    const simulation = d3.forceSimulation().nodes(nodesData);
+
+    // add forces
+    simulation
+      // charge force treats each node as a charged particle,
+      // so that nodes will repel each other when they get too close together
+      .force('charge_force', d3.forceManyBody())
+      // centering force drives all nodes torward the center of the canvas
+      .force('center_force', d3.forceCenter(width/2, height/2))
+      // link force constrains the nodes' motion based on how they are linked
+      .force('links', d3.forceLink(linksData).id(function(d) {
+        return d.hostname;
+      }));
+
+    // separate first from third party nodes
+    const nodesFirstParty = [];
+    const nodesThirdParty = [];
+
+    for (let i = 0; i < nodesData.length; i++) {
+      if (nodesData[i].party === 'first') {
+        nodesFirstParty.push(nodesData[i]);
+      } else {
+        nodesData[i].type = d3.symbolTriangle;
+        nodesData[i].size = 100;
+        nodesThirdParty.push(nodesData[i]);
       }
+    }
+
+    // draw circles for the firstParty nodes
+    const nodeFirstParty = svg.append('g')
+      .attr('class', 'nodes')
+      .selectAll('circle')
+      .data(nodesFirstParty)
+      .enter()
+      .append('circle')
+      .attr('r', 5)
+      .attr('fill', 'white');
+
+    // draw triangles for the thirdParty nodes
+    const nodeThirdParty = svg.append('g')
+      .attr('class', 'nodes')
```

Comment:
I agree it makes things very confusing, as each method returns a new value that the subsequent method manipulates further. These are D3 methods though, and it looks like they designed their methods to be used in this chain fashion. Given that the value of `nodeThirdParty` is changed with each method, wouldn't I need to use `let` instead of `const` and have several re-assignment statements following the declaration?

```javascript
let nodeThirdParty = svg.append('g');
nodeThirdParty = nodeThirdParty.attr('class', 'nodes');
// ...
```
Why is this better in your view?